src/gluonts/mx/trainer/callback.py

```diff
@@ -0,0 +1,541 @@
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Standard library imports
+from typing import List, Optional
+import logging
+
+# Third-party imports
+import numpy as np
+import mxnet.gluon.nn as nn
+import mxnet as mx
+from gluonts.dataset.common import Dataset
+from gluonts.mx.trainer.model_averaging import AveragingStrategy
+from gluonts.mx.trainer.model_iteration_averaging import (
+    IterationAveragingStrategy,
+)
+from mxnet import gluon
+
+# First-party imports
+from gluonts.evaluation import Evaluator
+from gluonts.mx.model.predictor import GluonPredictor
+from gluonts.mx.trainer.learning_rate_scheduler import MetricAttentiveScheduler
+from gluonts.support.util import copy_parameters
+
+# TODO: documentation
+
+
+class Callback:
+    """
+    Abstract Callback base class.
+    Callbacks control the training of the GluonTS trainer.
+    To write a custom Callback you can subclass Callback and overwrite one or more of the hook methods.
+    Hook methods with boolean return value stop the training if False is returned.
+    """
+
+    def __init__(self, **kwargs):
+        pass
+
+    def on_network_initializing_end(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        pass
+
+    def on_train_batch_start(self, training_network: nn.HybridBlock) -> None:
+        pass
+
+    def on_validation_batch_start(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        pass
+
+    def on_train_batch_end(self, training_network: nn.HybridBlock) -> None:
+        pass
+
+    def on_train_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        return True
+
+    def on_validation_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        return True
+
+    def on_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        return True
+
+    def on_train_end(
+        self,
+        training_network: nn.HybridBlock,
+        temporary_file: str,
+        ctx: Optional[mx.context.Context] = None,
+    ) -> None:
+        pass
+
+
+class CallbackList(Callback):
+    """
+    Used to chain multiple callbacks to one callback.
+    Boolean hook methods are logically joined with AND, meaning that if at least one callback method returns False, the training is stopped.
+
+    Parameters
+    ----------
+    callbacks
+        A list of gluonts.mx.trainer.callback.Callback's.
+    """
+
+    def __init__(self, callbacks: List[Callback], **kwargs):
```

Comment:
I think you will need to add the `@validated()` decorator to the constructor of all these classes, so that they are automatically serializable. You can check in the `Trainer` class how this is imported/used.